kernel/rros/observable.rs

```diff
@@ -0,0 +1,970 @@
+use alloc::{rc::Rc, sync::Arc};
+
+use crate::{
+    clock::{rros_ktime_monotonic, RrosClock, RROS_MONO_CLOCK},
+    factory::{
+        self, rros_element_is_observable, CloneData, RrosElement, RrosFactory, RrosFactoryInside,
+        RROS_CLONE_MASTER, RROS_CLONE_OBSERVABLE, RROS_OBSERVABLE_CLONE_FLAGS,
+    },
+    file::RrosFileBinding,
+    sched::{rros_schedule, RrosPollHead},
+    thread::{rros_init_user_element, CONFIG_RROS_NR_THREADS},
+    wait::{RrosWaitQueue, RROS_WAIT_PRIO},
+};
+use core::{cell::RefCell, default::Default, mem::size_of};
+use kernel::{
+    c_types, device, dovetail,
+    file::File,
+    file_operations::{FileOpener, FileOperations, IoctlCommand},
+    io_buffer::{IoBufferReader, IoBufferWriter},
+    irq_work::IrqWork,
+    ktime::{self, KtimeT},
+    linked_list::{GetLinks, Links, List},
+    prelude::*,
+    premmpt::running_inband,
+    rbtree, spinlock_init,
+    str::CStr,
+    sync::{HardSpinlock, Lock, SpinLock},
+    task,
+    uidgid::{KgidT, KuidT},
+    user_ptr::UserSlicePtr,
+    waitqueue,
+};
+
+pub const CONFIG_RROS_NR_OBSERVABLE: usize = 16;
+pub const RROS_OBSERVABLE_IOCBASE: char = 'o';
+pub const RROS_OBSIOC_SUBSCRIBE: u32 =
+    kernel::ioctl::_IOW::<RrosSubscription>(RROS_OBSERVABLE_IOCBASE as u32, 0);
+// pub const RROS_OBSIOC_UNSUBSCRIBE: u32 = kernel::ioctl::_IO(RROS_OBSERVABLE_IOCBASE as u32, 1);
+pub const RROS_NOTIFY_ONCHANGE: i32 = 1 << 0;
+pub const RROS_NOTIFY_MASK: i32 = RROS_NOTIFY_ONCHANGE;
+// pub const RROS_NOTIFY_ALWAYS: i32 = 0 << 0;
+pub const RROS_NOTIFY_INITIAL: i32 = 1 << 31;
+
+// Notice tags below this value are reserved to the core.
+pub const RROS_NOTICE_USER: u32 = 64;
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub union ObservableValue {
+    val: i32,
+    lval: i64,
+    ptr: *mut c_types::c_void,
+}
+
+impl ObservableValue {
+    fn new() -> Self {
+        Self { lval: 0 }
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct RrosNotice {
+    pub tag: u32,
+    pub event: ObservableValue,
+}
+
+impl RrosNotice {
+    fn new() -> Self {
+        Self {
+            tag: 0,
+            event: ObservableValue::new(),
+        }
+    }
+}
+
+pub struct RrosSubscriber {
+    pub subscriptions: SpinLock<rbtree::RBTree<u32, Arc<RrosObserver>>>,
+}
+
+impl RrosSubscriber {
+    pub fn new() -> Self {
+        Self {
+            subscriptions: unsafe { SpinLock::new(rbtree::RBTree::new()) },
+        }
+    }
+
+    pub fn init(&mut self) {
+        spinlock_init!(
+            unsafe { Pin::new_unchecked(&mut self.subscriptions) },
+            "RrosSubscriber"
+        );
+    }
```

Comment:
Why split these two functions? You can check the `new_and_init` function of the chrdev in RFL.

Reply:
Ok. I will merge them into one `new_and_init` function.